parser/src/main/java/com/linkedin/avroutil1/parser/avsc/AvscParser.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2021 LinkedIn Corp.
+ * Licensed under the BSD 2-Clause License (the "License").
+ * See License in the project root for license information.
+ */
+
+package com.linkedin.avroutil1.parser.avsc;
+
+import com.linkedin.avroutil1.model.AvroRecordSchema;
+import com.linkedin.avroutil1.model.AvroSchema;
+import com.linkedin.avroutil1.model.AvroType;
+import com.linkedin.avroutil1.parser.AvscParseContext;
+import com.linkedin.avroutil1.parser.Located;
+import com.linkedin.avroutil1.parser.exceptions.AvroSyntaxException;
+import com.linkedin.avroutil1.parser.exceptions.JsonParseException;
+import com.linkedin.avroutil1.parser.exceptions.ParseException;
+import com.linkedin.avroutil1.parser.jsonpext.JsonObjectExt;
+import com.linkedin.avroutil1.parser.jsonpext.JsonReaderExt;
+import com.linkedin.avroutil1.parser.jsonpext.JsonReaderWithLocations;
+import com.linkedin.avroutil1.parser.jsonpext.JsonStringExt;
+import com.linkedin.avroutil1.parser.jsonpext.JsonValueExt;
+import com.linkedin.avroutil1.util.Util;
+import jakarta.json.JsonValue;
+import jakarta.json.stream.JsonParsingException;
+
+import java.io.StringReader;
+import java.util.function.Supplier;
+
+/**
+ * parses avro schemas out of input assumed to be avsc.
+ * for the avsc specification see https://avro.apache.org/docs/current/spec.html#schemas
+ */
+public class AvscParser {
+
+    public AvscParseResult parse(String avsc) {
+        JsonReaderExt jsonReader = new JsonReaderWithLocations(new StringReader(avsc), null);
+        JsonObjectExt root;
+        AvscParseContext context = new AvscParseContext(avsc);
+        AvscParseResult result = new AvscParseResult();
+        try {
+            root = jsonReader.readObject();
+        } catch (JsonParsingException e) {
+            result.recordError(new JsonParseException("json parse error at " + e.getLocation(), e));
+            return result;
+        } catch (Exception e) {
+            result.recordError(new JsonParseException("unknown json parse error", e));
+            return result;
+        }
+
+        try {
+            parseSchemaDeclOrRef(root, context, true);
+            result.recordParseComplete(context);
+        } catch (Exception parseIssue) {
+            result.recordError(parseIssue);
+        }
+
+        return result;
+    }
+
+    /**
+     * parses a schema (or a reference to one by FQCN) out of a given json node
+     * @param node json node (which is a schema declaration or reference)
+     * @param context context for this parse operation
+     * @param topLevel true if this schema is the "top-level" schema for this context
+     *                 (for example outer-most schema in an avsc file)
+     * @throws ParseException
+     */
+    private void parseSchemaDeclOrRef (
```

Comment:
correct me if I am wrong: as of this PR, the parser does one pass of parsing the avsc file and populating the parse context object (of course, some AvroType parsing is still pending). 
Is there a second pass or what are the next steps here?

Reply:
in this PR is doesnt even parse fields, and so does not recurse into nested schemas.
in the future there's the whole recursive aspect, followed by some sort of 2nd pass to "stitch" references to external schemas. there will need to be a "cross-file" layer on top of AvscParseResult

Reply:
ok. this is where an RB description would help :) I can avoid asking annoying questions.